### DISCUSSION TITLE
Nested useSnapshot can fail a consistency check, triggering duplicate render invocations

### DISCUSSION THREAD

**Discussion by endash** [pmndrs/valtio#890] [Bug report] [answered]
The scenario here is pretty simple. Two components each using `useSnapshot`, one a child of the other, consuming _separate_ values from the store.

```
> Parent // relies on `a`
  -> Child // relies on `b`
```

The observed behavior when `a` changes is for `Parent` to rerender once, and then `Child` re-renders _twice_. (Note: this is _without_ strict mode.)

I've traced the internal behavior fairly closely, and my understanding of what is going on is:

1. When the store changes, both components' `getSnapshot`s are invoked outside-of-render. `Child` avoids a _direct_ re-render by checking the latest snapshot against the cached `lastSnapshot` and `lastAffected`. Since it doesn't rely on `a`, the only change, it returns the (now actually stale) cached `lastSnapshot`. `Parent` _does_ rely on `a`, so this check causes it to instead return the new `nextSnapshot`, and this difference triggers a re-render of `Parent`.

2. When invoked inside the render function, `getSnapshot` _always_ returns `nextSnapshot`.

3. This fresh snapshot will, eventually, be cached as `lastSnapshot` in a `useEffect` hook.

4. Before that `useEffect` happens for `Parent`, its child `Child` is rendered.

5. `getSnapshot` inside the `Child` render now returns `nextSnapshot`. This is different from the last call to `Child`'s `getSnapshot` and from the currently cached `lastSnapshot`. As above, `lastSnapshot` will be updated with this genuinely fresh snapshot in a `useEffect`.

6. `Child`'s render function returns, and the tree is finished rendering. `useEffect`s have _not yet been run_.

7. At this point, `Child`'s `getSnapshot` is invoked one more time, inside the React work loop, I believe by `isRenderConsistentWithExternalStores`. This is a consistency check against the possibility that the store might have changed since `getSnapshot` was invoked inside the render. The consistency check _fails_.

8. Why does it fail? Now that `Child`'s `getSnapshot` is being invoked outside-of-render, its behavior once again depends on whether its actual data dependencies have changed. Since it only relies on `b`, and `b` has _not_ changed, it reverts to returning its cached `lastSnapshot`. The cached `lastSnapshot` has _not_ yet been updated to reflect the `nextSnapshot` returned during the render.

9. `Child`'s `useEffect` hook is now invoked, updating its cached `lastSnapshot`.

10. The same consistency check for `Parent` is now run, and this one _passes_, because `Parent`'s dependency on `a` bypasses the (still stale) cache.

11. `Parent`'s `useEffect` hook is now invoked, updating its cached `lastSnapshot`. `Parent` re-render is now 100% complete.

11. `Child` is now re-rendered. This is a weird render, though. `useSnapshot` seems to behave normally, but `useEffect`s don't run, children don't re-render, and there's no additional consistency check. My tentative guess is that once the two in-render `getSnapshot` calls turn out to be consistent after all that the fiber just bails out.

### Attempted resolution

Changing the `useEffect` hook inside `useSnapshot` to a `useLayoutEffect` solves the timing issue, since `useLayoutEffect`s are run before the consistency check, and once the cached `lastSnapshot` is updated there's no inconsistency. `useLayoutEffect`s of course aren't run when rendering server-side, so I assume that is not a general solution.

**Comment by endash** [comment]
Upon further experimentation the nestedness of the `useSnapshot` hooks was a red herring, and the same consistency check failure happens to an isolated use, when the component re-renders anytime after a store mutation (of data that the component does not depend on.)

Since an even simpler replication turns out to be possible, I have trouble imagining that someone else hasn't stumbled across this before so I'll re-review existing issues and discussions to see if this is a known issue.

**Comment by dai-shi** [comment]
> Changing the `useEffect` hook inside `useSnapshot` to a `useLayoutEffect` solves the timing issue, since `useLayoutEffect`s are run before the consistency check, and once the cached `lastSnapshot` is updated there's no inconsistency.

Hm, that sounds like a reasonable fix. I actually thought that uLE should be appropriate, but at that time the use of uLE is warned in SSR, so avoided.

> `useLayoutEffect`s of course aren't run when rendering server-side, so I assume that is not a general solution.

Neither does useEffect.

----

Would you be able to create a failing test case?

> the same consistency check failure happens to an isolated use

So, do you mean it fails even without nesting?

**Reply by endash** [reply]
> Neither does useEffect.

Indeed. Not sure why I thought they were different, although it probably had to do with that warning. In that case, the change to uLE is probably the correct remediation.

> Would you be able to create a failing test case?

Should be simple enough. Will try to pull that together when I get a minute.

> So, do you mean it fails even without nesting?

Yeah. I initially thought the nesting was significant (and that's how I first encountered this in my real app, with most state going through the store) but all that actually ended up mattering on closer inspection is the delay in updating the cached snapshot because of the `useEffect` timing relative to the consistency check.

**Reply by endash** [reply]
Actually, once I could replicate without nesting I had realized last night that the much simpler scenario I was looking at was probably already under test (and it does seem to be, as a "basic" test to boot) so there's likely some subtlety that is currently escaping me. I'll update this either when I can get a failing test case or if I end up throwing my hands up.

**Reply by endash** [reply]
At the risk of driving myself a bit crazy with the nesting question, a straightforward duplication of the existing test against extra render function calls DOES demonstrate the bug when the two components are nested, so that's what I've gone with for a first stab. The failing test passes once `useEffect` is replaced by `useLayoutEffect`.

https://github.com/endash/valtio/commit/d9ece850d0f1b3127bc8a7202464b913e655025e

**Reply by endash** [reply]
OK, it turned out to be straightforward enough to demonstrate the bug without nesting `useSnapshot`, although there's still a nesting component structure so that the `Parent` re-rendering is what triggers the `Child` (with the `useSnapshot`) to re-render. 

Both failing tests pass with `useEffectLayout`. The second test also passes if line 271 is commented out, demonstrating that the extra render is still caused in part by a change to the store.

https://github.com/endash/valtio/commit/d67f6561c5aeb0db109dc3eda4fa7558081477c1

**Reply by dai-shi** [reply]
Thanks so much!

Yeah, it makes sense.
Now, I think we have two options:
- Live with the current behavior
- Fix with useLayoutEffect

I personally would like to fix, but it will show false-positive warning on SSR with old React (which we claim to support), which is very annoying and I'm sure people will start complaining (because there is no way to opt-out).

Hm, I have another idea:
- Make an opt-in option to use `useLayoutEffect` (Hack implementation)

**Reply by endash** [reply]
Supported versions of React that still have the warning are a problem. My preference would be to allow those users to opt-_out_ of `useLayoutEffect`, rather than for newer React users to have to opt-_in_. If anyone's still around maintaining those code bases enough to be annoyed by the warning they could address it with the the opt-out.

An open question is how much of a performance penalty these extra renders are in practice. A second question is whether they might be introducing any other runtime bugs. In theory that shouldn't be the case, since the render functions themselves ought to be idempotent, but there are probably edge cases out there. As for the first question, my hunch is that these extra renders must be happening quite a lot in any reasonably complex app, and even idempotent shallow renders can be expensive in many cases. I noticed the bug while tightening up some uses of `useMemo` for things I knew were very expensive, but my approach in general is to avoid `useMemo` for the sake of reasonableness, and I can see how that might come back to bite with behavior like this. (My current project is definitely on the small side overall, though, and so isn't a very good test case for investigating these questions.)

What I _personally_ would do is just fix it straight-up and then wait for the complaints, then explain that it addresses extra renders (are those not every React dev's greatest enemy??) and see who is actually so wedded to their old React version.

**Reply by dai-shi** [reply]
We are about to release v2 #810, probably right after React 19 is out (not sure when).
So, if we just fix it and wait for complaints, let's do it in v2.

**Reply by endash** [reply]
That's probably the right approach 👍

**Reply by dai-shi** [reply]
Cool. Can you send a PR to `v2` branch please?

**Reply by endash** [reply, marked as answer]
Done. Thanks for being so responsive!

**Reply by dai-shi** [reply]
PR 👉 #891

Published: https://www.npmjs.com/package/valtio/v/2.0.0-beta.5
